Fix accuracy for layernorm/rmsnorm benchmarking

shunting314 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #166005

Example command:
    python benchmarks/dynamo/genai_layers/benchmark.py --exit-on-accuracy-failure --tolerance=1e-2 rmsnorm_backward

Fix the accuracy problem for layernorm/rmsnorm fwd/bwd.
Also fix some quack calls (maybe due to quack API change)

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames @Lucaskabela

shunting314 committed:
Fix accuracy for layernorm/rmsnorm benchmarking

Example command:
    python benchmarks/dynamo/genai_layers/benchmark.py --exit-on-accuracy-failure --tolerance=1e-2 rmsnorm_backward

Fix the accuracy problem for layernorm/rmsnorm fwd/bwd.
Also fix some quack calls (maybe due to quack API change)

[ghstack-poisoned]

BoyuanFeng commented on benchmarks/dynamo/genai_layers/benchmark.py:
```diff
@@ -72,23 +73,25 @@ def run_benchmark(
 
     benchmark_class = BENCHMARK_REGISTRY[benchmark_name]
-    benchmark = benchmark_class(compile_mode)
+    benchmark = benchmark_class(args, compile_mode)
```
maybe rename `args` to `bench_settings`? `args` sound like tensor inputs.

BoyuanFeng commented on benchmarks/dynamo/genai_layers/kernels.py:
```diff
@@ -614,20 +662,31 @@ def compiled(self, args, kwargs=None) -> Any:
 
         x, w, dy = args
+        eps = 1e-6
+        mean, rstd = self.compute_mean_rstd(x, eps)
```
did liger change the api so we need another function `compute_mean_rstd`?

shunting314 replied:
Liger save mean/rstd in low precision and cause accuracy problem. Check the docstring above for more details

BoyuanFeng commented on benchmarks/dynamo/genai_layers/kernels.py:
```diff
@@ -454,24 +474,52 @@ def compiled(self, args, kwargs=None) -> Any:
+                dresidual=None,
+                sm_count=sm_count,
+            )
+            dw = dw_partial.sum(dim=0).to(w.dtype)
```
`dw = dw_partial.sum(dim=0).to(w.dtype)` is not fused with `_rmsnorm_bwd` in eager. I am not sure if this still represents quack performance..

shunting314 committed:
Update on "Fix accuracy for layernorm/rmsnorm benchmarking"


Example command:
    python benchmarks/dynamo/genai_layers/benchmark.py --exit-on-accuracy-failure --tolerance=1e-2 rmsnorm_backward

Fix the accuracy problem for layernorm/rmsnorm fwd/bwd.
Also fix some quack calls (maybe due to quack API change)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

shunting314 commented:
@pytorchbot merge
